### ISSUE TITLE
Requirements for WebID document server

### ISSUE THREAD

**Issue by RubenVerborgh** [solid/specification#376]
The Solid Protocol does not describe the requirements for the HTTP server hosting a WebID document.

In particular, we currently have:

> When a WebID is dereferenced, server provides a representation of the WebID Profile in an RDF document

but there is no constraint in our document on what RDF representations should be available.

The [linked WebID spec](https://www.w3.org/2005/Incubator/webid/spec/identity/) (not under control of the Solid Community Group) says:

> A WebID Profile Document is a Web resource that MUST be available as text/turtle [turtle], but MAY be available in other RDF serialization formats (e.g. [RDFA-CORE]) if requested through content negotiation. 

This contrasts with the constraints on Solid resource server:

> the server MUST accept GET requests on this resource when the value of the Accept header requests a representation in text/turtle or application/ld+json

---

I have seen applications assume that similar requirements hold for WebID document servers, i.e., that Turtle or JSON-LD can be assumed (or even a specific JSON-LD frame). However, this is not the case.

Other Solid specifications seem to also (implicitly) make assumptions, notably:

#### Solid-OIDC
> The user’s WebID in the webid claim MUST be dereferenced […] the AS MUST check the WebID document for the existence of a statement matching ?webid <http://www.w3.org/ns/solid/terms#oidcIssuer> ?iss

#### Web Access Control
> Servers are strongly discouraged from trusting the information returned by looking up an agent’s WebID for access control purposes.

---

We probably want to be explicit here.

// Reported based on comments from @ThisIsMissEm and @joachimvh

**Comment by matthieubosquet** [comment]
I think the problem here is that the Solid Protocol says too much about WebID. It makes sense to me that a reference is made to it, but the middle statement should probably go:

> When a WebID is dereferenced, server provides a representation of the WebID Profile in an RDF document [RDF11-CONCEPTS] which uniquely describes an agent denoted by a WebID.

Without that statement above, the spec could link to the relevant documents, respecting the [spec orthogonality](https://www.w3.org/blog/2009/06/orthogonality-of-specification/) principle as such:

> A WebID is an HTTP URI denoting an agent, for example a person, organisation, or software [WEBID].
>
> WebIDs are an underpinning component in the Solid ecosystem and are used as the primary identifier for users and applications [SOLID-OIDC] [WEBID-TLS].

**Comment by csarven** [comment]
>The Solid Protocol does not describe the requirements for the HTTP server hosting a WebID document.

[1] The Solid Protocol does describe a WebID document served from a Solid server, i.e., to have representations of a WebID Profile Document available in Turtle and JSON-LD, and as per content negotiation it may be available in other media types (typically in concrete RDF syntaxes).

>Other Solid specifications seem to also (implicitly) make assumptions, notably:
>Servers are strongly discouraged from trusting the information returned by looking up an agent’s WebID for access control purposes.

Not relevant because WAC does not have a requirement to dereference a WebID or have the WebID Profile Document available in any particular format. WebID is only used as an identifier for agents.

>I think the problem here is that the Solid Protocol says too much about WebID. It makes sense to me that a reference is made to it, but the middle statement should probably go:

The statement is in context of the Solid Protocol [1] - it may be redundant and can be removed on that basis.

---

The WebID section can be updated to be more clear about *any* "HTTP server hosting a WebID document" as opposed to strictly Solid servers as per the Solid Protocol. This is essentially factoring in the requirements put forward by the other core specifications. The current text using "RDF document" in the normative section generally covers that possibility:

>When a WebID is dereferenced, server provides a representation of the WebID Profile in an RDF document [RDF11-CONCEPTS] which uniquely describes an agent denoted by a WebID.

It leaves it up to each specification in the ecosystem to make their own specific and definitive calls.

If instead all of the Solid specifications want to use one-requirement-to-rule-all-requirements on WebID in the whole ecosystem as per Solid Protocol, we can also do that - provided that the groups working on each specification are aligned with that. (As it stands, that practically means acknowledging both Turtle and JSON-LD.)

**Comment by RubenVerborgh** [comment]
> > The Solid Protocol does not describe the requirements for the HTTP server hosting a WebID document.
> 
> [1] The Solid Protocol does describe a WebID document served from a Solid server

Do you have the exact phrasing? I cannot find it (search term: `WebID`).

**Comment by csarven** [comment]
As per WebID 1.0 ED, a WebID document is an RDF document. Representation in Turtle being the minimum requirement for interop whilst standard content negotiation takes place.

The Solid Protocol specifies Solid server requirements for RDF documents: https://solidproject.org/TR/2021/protocol-20211207#server-representation-turtle-jsonld . A conforming Solid server will provide WebID documents in Turtle and JSON-LD.

Currently, the Solid Protocol is not describing how any HTTP server (under the sun) should serve a WebID document (besides it being an RDF document and adhering to WebID 1.0 ED). The Solid Protocol does not - or in a position to - dictate how WebID documents are to be served from HTTP servers besides the Solid server or potentially the servers that can participate in the Solid ecosystem.

As said earlier, in order to use the Solid Protocol as a reference point for all specifications in the Solid ecosystem, there needs to be agreement across the current (and upcoming) specifications on a requirement that they can all rely on. In its absence, the current WebID section in the Solid Protocol merely states that those documents are deemed to be RDF documents. Hence, a server hosting a WebID document in the Solid ecosystem is a Solid server, in which the requirements are given by the Solid Protocol (Turtle, JSON-LD) or a server (as defined by a particular specification in the Solid ecosystem) that conforms to WebID 1.0 ED (Turtle). As it stands, any other kind of "server" would either be not conforming to WebID 1.0 ED or Solid Protocol or extends either one of those.

**Comment by RubenVerborgh** [comment]
Thanks; that is sufficient for me. Can be closed.

**Comment by timbl** [comment]
This is NOT clear IMHO because we have a WebID 1.0 Editors Draft spec which specifies what the server of a WebID must support, and the SolidOIDC people use the term "WebID" without distancing themselves from the WebID spec, and not supporting it.

**Comment by csarven** [comment]
Specifications defining roles that serve a WebID Profile Document that do not conform the Solid server should specify the ways that the consumer of the WebID Profile Document can discover the resource which accepts read-write operations as per the Solid protocol. For example, Solid WebID Profile spec can describe the property for its discovery, e.g., `solid:profileDocument`. The other specs, e.g., Solid-OIDC can recommend its use.

**Comment by timbl** [comment]
You are saying that SolidOIDC ID document must contain something like
`<#me>  solid:profileDocument <https://example.com/aliice/card> .`
then ? 
or maybe 
`<#me>  solid:ID. <https://example.com/aliice/card#me> .` .... ?

**Comment by elf-pavlik** [comment]
AFAIK all comments about `solid:profileDocument` are about 
Solid WebID Profile:  https://solid.github.io/webid-profile/
NOT about Solid-OIDC: https://solidproject.org/TR/oidc#webid-profile

**Comment by csarven** [comment]
Tim, with respect to this issue, I'm saying that any specification that requires or anticipates dereferencing WebIDs - which may resolve to a non-Solid server - can consider recommending the discovery of that person's other or preferred WebID that is hosted on a Solid server.

My inquiry about solid:profileDocument or solid:webid was about what's intended to be discovered. I think solid:webid makes more sense as that actually seems to be what's intended - enables any WebID (on the face of the planet) to link to a Solid WebID where a Solid-aware application can perform read-write operations. I've created https://github.com/solid/webid-profile/issues/58 to specifically discuss that concept.

I recommend closing this issue with the understanding that Solid Protocol is clear about the requirements. If other specs have different requirements, and want to make it possible for applications to find their way to WebID that in conforms to the Solid Protocol, there needs to be a discovery mechanism - a property perhaps.

**Comment by csarven** [comment]
Changed the status to Waiting for Commenter again and asking @RubenVerborgh whether the answers are satisfactory and asking if @timbl is satisfied or has objections to close this particular issue.

**Comment by RubenVerborgh** [comment]
> I recommend closing this issue with the understanding that Solid Protocol is clear about the requirements.

I'm afraid I disagree upon looking more closely.

I find in the spec:

> When a server creates a resource on HTTP PUT, POST or PATCH requests such that the request’s representation data encodes an RDF document [[RDF11-CONCEPTS](https://solidproject.org/TR/2021/protocol-20211217#bib-rdf11-concepts)] (as determined by the Content-Type header), the server MUST accept GET requests on this resource when the value of the Accept header requests a representation in text/turtle or application/ld+json [[Turtle](https://solidproject.org/TR/2021/protocol-20211217#bib-turtle)] [[JSON-LD11](https://solidproject.org/TR/2021/protocol-20211217#bib-json-ld11)].

—https://solidproject.org/TR/2021/protocol-20211217#server-representation-turtle-jsonld

> A WebID is an HTTP URI denoting an agent, for example a person, organisation, or software [[WEBID](https://solidproject.org/TR/2021/protocol-20211217#bib-webid)]. When a WebID is dereferenced, server provides a representation of the WebID Profile in an RDF document [[RDF11-CONCEPTS](https://solidproject.org/TR/2021/protocol-20211217#bib-rdf11-concepts)] which uniquely describes an agent denoted by a WebID. WebIDs are an underpinning component in the Solid ecosystem and are used as the primary identifier for users and applications.

—https://solidproject.org/TR/2021/protocol-20211217#webid

(Side-note: this paragraph has chosen not to use MUST language.)

Whereas @csarven writes above

> [1] The Solid Protocol does describe a WebID document served from a Solid server,

I just don't see it in the spec; I don't see a line saying "WebIDs MUST be served from a Solid server".
I only see the rather ambiguous "When a WebID is dereferenced, server _[sic]_ provides a representation of the WebID Profile in an RDF document", but there is no specific mention of this being a Solid server.

I.e., if these above lines are sufficient, could someone please ELI5 me step by step how I know that the WebID MUST be on a Solid server; and if that is too much of an inference to make, can't we just add it explicitly?

**Comment by csarven** [comment]
I feel we are now awfully close to repeating ourselves. For example, the last paragraph in https://github.com/solid/specification/issues/376#issuecomment-1022088555 as to what the current expectations are and whether people want something stricter, e.g., make sure all Solid specs work with or set the same requirements for WebID representations.

>(Side-note: this paragraph has chosen not to use MUST language.)
>I only see the rather ambiguous "When a WebID is dereferenced, server [sic] provides a representation of the WebID Profile in an RDF document", but there is no specific mention of this being a Solid server.

So? It is a normative section. It is describing something about the ecosystem where the Solid Protocol plays one of the key roles. The Solid Protocol is only describing something that is already compatible with the current WebID 1.0 ED.

>I just don't see it in the spec; I don't see a line saying "WebIDs MUST be served from a Solid server".

>I.e., if these above lines are sufficient, could someone please ELI5 me step by step how I know that the WebID MUST be on a Solid server; and if that is too much of an inference to make, can't we just add it explicitly?

Different things are referencing WebIDs and different things are dereferencing WebIDs.

A Solid application needing to read the WebID Profile Document (from somewhere on the Web) only needs to follow (current) WebID 1.0 ED's requirement, i.e., Turtle or another concrete RDF syntax depending on negotiation. A Solid application will need to be equipped to work with Turtle or JSON-LD so that it can communicate with a Solid server. If a subset of the WebIDs that the application wants somewhere on the Web is on a Solid server, all is still well.

I'd argue - anecdotal evidence - that we've been building applications with that understanding, i.e., Solid applications do read WebIDs that are not hosted on Solid servers. (For example, and as you know, dokieli will go through all of my and my contact's connected WebID when it needs to. The use case is to read. It doesn't have a need to write to every WebID based on Solid) There is a lot of value to be able to use WebIDs that are not strictly on a Solid server. There'll be more applications needing to read than to write WebID Profile Documents.

Having said that, I'd still argue that we need to strive for a mutual understanding that when a WebID is not a Solid server, then obviously a Solid application won't be able to perform (access controlled) write requests to it. And, for those subset of WebID owners (one having sufficient rights to update the WebID Profile Document by any means - see e.g., webid-profile issue 58) wanting to lead Solid applications to follow their nose to their equivalent or alternate WebIDs that are on a Solid server, then we specify that bridge.

Hope that helps as to at least how I'm looking at this.

**Comment by langsamu** [comment]
Somewhat related implementation feedback, though not addressing the substantial question.

In the Inrupt Enterprise Solid Server 2.0 (which I contribute to) **WebIDs are not stored on a Solid server**. Instead they are stored on a separate service that adheres to WebID ED and Solid-OIDC.

This is to support multiple use cases and customer requirements regarding separation of identity from storage.
For example identity in the form of WebIDs might be provided by someone other then the storage provider.
Another example: Access control to the WebID profile might fundamentally differ from access control to Solid resources.

It would be impossible to support these use cases if the WebID of the Pod owner was required to be stored in the Pod they own.
And it would be too much effort for implementers if we just required WebID servers to be Solid servers.

**Comment by woutermont** [comment]
@RubenVerborgh, if something still bothers you about this, could you try to explain it once more? A few days ago I thought I was following, but I'm no longer sure of that. 

I.m.o. the interrelations are quite clear: Solid specifies servers, WebID specifies identifiers and documents; Solid servers can host non-WebID resources, non-Solid servers can host valid WebID documents, and a WebID document on a Solid server adheres to both. Most importantly, for most of the very specific cases where WebIDs are used, one hosted outside of Solid works perfectly. Nothing in the spec seems to contradict that, even if some wording could be clearer.

I do agree with @matthieubosquet, however, that the Solid protocol (and other Solid specs) contains too much useless text about WebIDs. Most of it is repetition or paraphrasing of the WebID spec, which not only makes it redundant, but also increases the risk of confusion should the wording be not exactly the same. Some of it also creates too tight a coupling with WebID, where a simpler description (e.g. "dereferencable to RDF resource containing X") would suffice.

**Comment by woutermont** [comment]
@csarven, I might have used some rhetorical hyperbole, the term 'useless' should probably have been more in the lines of 'redundant', and once more our keenness for abstraction seems to differ. The point remains: there is text that would not matter if left out, and if the instability of a draft is the issue, then the best way would be a referrence to its fixed version. 

I do not see what about this is less constructive than @TallTed's contribution, and I feel in fact somewhat hurt in your karikature of me. Especially since my original comment just asked for confirmation that this issue could be closed again.

**Comment by melvincarvalho** [comment]
>  I feel in fact somewhat hurt in your karikature of me

@csarven please, be respectful.  Try to phase feedback, without something that could be interpreted as a passive aggressive jab.  Let's assume good faith, and that everyone is trying to be constructive.  Thanks!

**Comment by bblfish** [comment]
Just my two cents. 

  * I am ok with clients being required to know JSon-LD and Turtle,
  * my recent inspection of json-ld parsers for Java gave me a bad impression: I looked at Titanium and it is not async, need to open a new connection (probably on a java thread), etc... That does not look good to me. If I had time I would sit down and fix that, but it could take two months. 
  * Aaron's use case for json-ld he put forward here seems a reasonable one https://github.com/w3c/WebID/issues/3 

So my feeling is for exceptional cases like that it makes sense.

**Comment by RubenVerborgh** [comment]
Thanks all, I think we have enough for now, and other issues to follow up on related discussions.
